src/magic_folder/web.py

```diff
@@ -114,11 +134,158 @@ class APIv1(Resource, object):
+    """
+    Implements the v1 API for the ``/participants/<magic-folder-name>``
+    part of the API hierarchy.
+    """
+    _folder_config = attr.ib()
+    _tahoe_client = attr.ib()
+
+    # XXX maybe we could/should pass around a TahoeClient with the
+    # global-config? It could maybe simplify stuff, *so long as* we
+    # can still override a "testing" client into e.g. a "testing"
+    # config...
+
+    def __attrs_post_init__(self):
+        Resource.__init__(self)
+
+    def render_GET(self, request):
+        """
+        List all participants of this folder
+        """
+        collective = participants_from_collective(
+            self._folder_config.collective_dircap,
+            self._folder_config.upload_dircap,
+            self._tahoe_client,
+        )
+        d = collective.list()
+
+        def listed(participants):
+            reply = {
+                part.name: {
+                    "personal_dmd": part.dircap.decode("ascii"),
+                    # not tracked properly yet
+                    # "public_key": part.verify_key.encode(Base32Encoder),
+                }
+                for part in participants
+            }
+            request.setResponseCode(http.OK)
+            _application_json(request)
+            request.write(json.dumps(reply).encode("utf8"))
+        d.addCallback(listed)
+
+        def failed(reason):
+            # probably should log this failure
+            request.setResponseCode(http.INTERNAL_SERVER_ERROR)
+            _application_json(request)
+            request.write(json.dumps({"reason": "unexpected error processing request"}))
+            return None
+        d.addErrback(failed)
+        d.addBoth(lambda ignored: request.finish())
+        return NOT_DONE_YET
+
+    def render_POST(self, request):
+        """
+        Add a new participant to this folder with details from the JSON-encoded body.
+        """
+        body = request.content.read()
+        try:
+            participant = json.loads(body)
+            required_keys = {
+                "author",
+                "personal_dmd",
+            }
+            required_author_keys = {
+                "name",
+                # not yet
+                # "public_key_base32",
+            }
+            if set(participant.keys()) != required_keys:
+                raise ValueError("Require input: {}".format(", ".join(sorted(required_keys))))
+            if set(participant["author"].keys()) != required_author_keys:
+                raise ValueError("'author' requires: {}".format(", ".join(sorted(required_author_keys))))
+
+            author = create_author(
+                participant["author"]["name"],
+                # we don't yet properly track keys but need one
+                # here .. this won't be correct, but we won't use
+                # it .. following code still only looks at the
+                # .name attribute
+                VerifyKey(os.urandom(32)),
```

Comment:
Can you file an issue for this and link it here?

Reply:
https://github.com/LeastAuthority/magic-folder/issues/331 already exists (but yeah linking it in the comment is a good idea).